contributors/design-proposals/federated-placement-policy.md

```diff
@@ -0,0 +1,271 @@
+
+### 1. Initial Placement Decision
+
+To provide policy-based control over the initial placement decision, we propose
+a new admission controller that integrates with OPA:
+
+When admitting requests, the admission controller executes an HTTP API call
+against OPA. The API call passes the JSON representation of the resource in the
+message body.
+
+The response from OPA contains the desired value for the resource’s annotations
+(defined in policy by the administrator). The admission controller updates the
+annotations on the resource and admits the request:
+
+![InitialPlacement](https://docs.google.com/drawings/d/1c9PBDwjJmdv_qVvPq0sQ8RVeZad91vAN1XT6K9Gz9k8/pub?w=812&h=288)
+
+The update is performed by **merging** the annotations in the response with
+existing annotations on the resource. If there are overlapping keys, the value
+defined by policy is chosen.
+
+#### Example Policy Engine Query:
+
+```http
+POST /v1/data/io/k8s/federation/annotations HTTP/1.1
+Content-Type: application/json
+```
+
+```json
+{
+  "input": {
+    "apiVersion": "extensions/v1beta1",
+    "kind": "ReplicaSet",
+    "metadata": {
+      "annotations": {
+        "eu-jurisdiction-required": "true",
+        "pci-compliance-level": "2"
+      },
+      "creationTimestamp": "2017-01-23T16:25:14Z",
+      "generation": 1,
+      "labels": {
+        "app": "nginx-eu"
+      },
+      "name": "nginx-eu",
+      "namespace": "default",
+      "resourceVersion": "364993",
+      "selfLink": "/apis/extensions/v1beta1/namespaces/default/replicasets/nginx-eu",
+      "uid": "84fab96d-e188-11e6-ac83-0a580a54020e"
+    },
+    "spec": {
+      "replicas": 4,
+      "selector": {...},
+      "template": {...},
+    }
+  }
+}
+```
+
+#### Example Policy Engine Response:
+
+```http
+HTTP/1.1 200 OK
+Content-Type: application/json
+```
+
+```json
+{
+  "result": {
+    "federation.kubernetes.io/replica-set-preferences": {
+      "clusters": {
+        "gce-europe-west1": {
+          "weight": 1
+        },
+        "gce-europe-west2": {
+          "weight": 1
+        }
+      },
+      "rebalance": true
+    }
+  }
+}
+```
+
+> This example shows the policy engine returning the replica-set-preferences.
+> The policy engine could similarly return a desired value for other annotations
+> such as the Cluster Selector annotation.
+
+#### Error Handling
+
+The admission controller is designed to **fail closed**. Request handling may
+fail because of:
+
+- Serialization errors
+- Request timeouts or other network errors
+- Unexpected errors from the policy engine
+
+In the event of request timeouts (or other network errors) or back-pressure
+hints from the policy engine, the admission controller should retry after
+applying a backoff.
+
+If the administrator has not defined a policy yet, the response from the policy
+engine indicates no annotations are defined. The admission controller does not
+treat this as an error, i.e., it admits the request.
+
```

Comment:
Yet another thought...

We have the admission controller enabled by default, however, the default behaviour is to fail-open. When handling a request, the admission controller would read an annotation on a resource representing the policy engine (e.g., the service or replicaset). The annotation would indicate whether the admission controller should fail-open or fail-closed.

This way, once the policy engine was deployed the annotation could be updated to fail-closed. This could be done manually by an admin or automatically by the policy engine itself.